src/io.rs

```diff
@@ -32,30 +39,323 @@ where
+    ///     for rec in records.iter() {
+    ///         writer.write_record(rec.as_bytes()).unwrap();
+    ///     }
+    /// }
+    /// let read = std::io::BufReader::new(Cursor::new(buf));
+    /// let mut reader = RecordReader::new(read);
+    /// let mut ary = [0u8; 4];
+    /// let mut i = 0;
+    /// loop {
+    ///     let next = reader.read_next(&mut ary);
+    ///     match next {
+    ///         Ok(res) => match res {
+    ///             Some(len) => assert_eq!(&ary[0..len], records[i].as_bytes()),
+    ///             None => break,
+    ///         },
+    ///         Err(RecordReadError::CorruptFile) | Err(RecordReadError::IoError { .. }) => {
+    ///             break;
+    ///         }
+    ///         _ => {}
+    ///     }
+    ///     i += 1;
+    /// }
+    /// ```
+    /// When we may need to dynamically resize our buffer, use this peek_next_len()
+    /// ```
+    /// use tensorflow::io::{RecordReadError, RecordReader, RecordWriter};
+    /// use std::{io::Cursor, rc::Rc};
+    /// let mut buf = Vec::new();
+    /// let mut rc = Rc::new(&mut buf);
+    /// let records = vec!["foo", "barr", "baz"];
+    /// {
+    ///     let mut writer = RecordWriter::new(Rc::get_mut(&mut rc).unwrap());
+    ///     for rec in records.iter() {
+    ///         writer.write_record(rec.as_bytes()).unwrap();
+    ///     }
+    /// }
+    /// let read = std::io::BufReader::new(Cursor::new(buf));
+    /// let mut reader = RecordReader::new(read);
+    /// let mut vec = Vec::new();
+    /// while let Ok(Some(len)) = reader.peek_next_len() {
+    ///     let len = len as usize;
+    ///     if vec.len() < len {
+    ///         vec.resize(len, 0);
+    ///     }
+    ///     let next = reader.read_next(&mut vec[0..len]);
+    ///     assert_eq!(next.unwrap().unwrap(), len);
+    ///     // &vec[0..len] contains the bytes of this record
+    /// }
+    /// assert_eq!(vec.len(), 4);
+    /// ```
+    ///
+    pub fn read_next(&mut self, buf: &mut [u8]) -> Result<Option<usize>, RecordReadError> {
+        let len = match self.read_next_len_unchecked()? {
+            Some(len) => len,
+            None => return Ok(None),
+        };
+        if (buf.len() as u64) < len {
+            self.reader.seek(SeekFrom::Current(8 + len as i64))?;
+            return Err(RecordReadError::BufferTooShort { needed: len });
+        }
+        let mut len_bytes = [0u8; 8];
+        LittleEndian::write_u64(&mut len_bytes, len);
+        let len_ok = self.checksum(&mut len_bytes)?;
+
+        let slice = &mut buf[0..len as usize];
+        self.read_bytes_exact_unchecked(slice)?;
+        let bytes_ok = self.checksum(slice)?;
+
+        if bytes_ok && len_ok {
+            return Ok(Some(len as usize));
+        } else {
+            if !len_ok {
+                return Err(RecordReadError::CorruptFile);
+            }
+            return Err(RecordReadError::CorruptRecord);
+        }
```

Comment:
This could be simplified to:

```
if !len_ok {
  Err(RecordReadError::CorruptFile)
} else if !bytes_ok {
  Err(RecordReadError::CorruptRecord)
} else {
  Ok(Some(len as usize))
}
```

Reply:
Heh. I think I was a little tired last night.